test/box-luatest/gh_8204_fast_offset_test.lua

```diff
@@ -0,0 +1,907 @@
+                               {3}, {3, 0}, {3, 1}, {3, 2}, {3, 3},
+                               {4}, {4, 0}, {4, 1}, {4, 2}, {4, 3}}
+
+        -- Prepare proxies.
+        local txn_proxy = require('test.box.lua.txn_proxy')
+        local tx = txn_proxy.new()
+        local tx1 = txn_proxy.new()
+        local tx2 = txn_proxy.new()
+
+        -- Proxy helpers.
+        local conflict = {{error = "Transaction has been aborted by conflict"}}
+        local success = ''
+
+        -- Stringify a value to use in lua code string.
+        -- E. g. array of 2 elements is transformed into "{1, 2}" string.
+        local function to_lua_code(value)
+            return require('luatest.pp').tostring(value)
+        end
+
+        -- Check if count with given key and iterator gives the expected result
+        -- for the given transaction.
+        local function check(tx, it, key, expected_count, file, line)
+            -- The location of the callee.
+            local file = file or debug.getinfo(2, 'S').source
+            local line = line or debug.getinfo(2, 'l').currentline
+
+            -- The stringified key.
+            local key = to_lua_code(key)
+
+            local code = string.format('box.space.test.index.pk:count(%s, ' ..
+                                       '{iterator = "%s"})', key, it)
+
+            local comment = string.format('\nkey: %s,\niterator: %s,' ..
+                                          '\nfile: %s,\nline: %d,',
+                                          key, it, file, line)
+
+            local ok, res = pcall(tx, code)
+            t.assert(ok, comment)
+            t.assert_equals(res, {expected_count}, comment)
+        end
+
+        -- Make the tx1 open a transaction and count by given it and key,
+        -- then make the tx2 insert/delete (op) the given tuple,
+        -- then make the tx1 writing to make it abort on conflict,
+        -- then make the tx1 commit its transaction and expect tx1_result.
+        --
+        -- The tuple inserted by tx2 is cleaned up, the "writer" space is not.
+        local function count_do(tx1, tx2, it, key, expected_count, op, tuple,
+                                tx1_result)
+            assert(op == 'insert' or op == 'delete' or op == 'replace')
+
+            local old_len = s:len()
+            local tuple_existed = s:count(tuple) ~= 0
+
+            -- The location of the callee.
+            local file = debug.getinfo(2, 'S').source
+            local line = debug.getinfo(2, 'l').currentline
+
+            -- Remove past stories cause they cause unconditional conflicts,
+            -- whereas future statements only conflict with count if they
+            -- insert a new matching tuple or delete a counted one.
+            box.internal.memtx_tx_gc(100)
+
+            tx1:begin()
+            check(tx1, it, key, expected_count, file, line);
+            tx2(string.format('box.space.test:%s(%s)', op, to_lua_code(tuple)))
```

Comment:
One should check if the operation succeed (replace actually replaced, insert actually inserted, etc.).